### PR TITLE
Add http server to Aws:Ec2

### PR DESCRIPTION
**Pull request by jonathangold**
Add an http server that will ultimately be used as an endpoint for amazon SNS.

### CODE REVIEW

File: resources/aws_ec2.go [hunk 1]
```diff
@@ -71,11 +74,12 @@ var AwsRegions = []string{
...
-	State   string `yaml:"state"`   // state: running, stopped, terminated
-	Region  string `yaml:"region"`  // region must match an element of AwsRegions
-	Type    string `yaml:"type"`    // type of ec2 instance, ie. t2.micro
-	ImageID string `yaml:"imageid"` // imageid must be available on the chosen region
+	BaseRes         `yaml:",inline"`
+	State           string `yaml:"state"`           // state: running, stopped, terminated
+	Region          string `yaml:"region"`          // region must match an element of AwsRegions
+	Type            string `yaml:"type"`            // type of ec2 instance, eg: t2.micro
+	ImageID         string `yaml:"imageid"`         // imageid must be available on the chosen region
+	WatchListenAddr string `yaml:"watchlistenaddr"` // the local address or port connected to the above url, eg: 10.0.0.0:23456 or 23456
```
Comment (purpleidea) [inline resources/aws_ec2.go]: what "above url" ?
(i realize this might come later, but yeah, maybe reword?)

File: resources/aws_ec2.go [hunk 2]
```diff
@@ -677,3 +681,40 @@ func (obj *AwsEc2Res) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+			case <-obj.closeChan:
+			}
+			return
+		}
+	}()
+	return srv
+}
+
+// postHandler listens for posts on the SNS Endpoint.
+func (obj *AwsEc2Res) postHandler(w http.ResponseWriter, req *http.Request) {
```
Comment (purpleidea) [inline resources/aws_ec2.go]: since this is just used by sns, maybe `snsPostHandler` ?

File: resources/aws_ec2.go [hunk 3]
```diff
@@ -677,3 +681,40 @@ func (obj *AwsEc2Res) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+	if req.Method == "POST" {
+		post, _ := ioutil.ReadAll(req.Body)
+		log.Printf(string(post))
+	} else {
```
Comment (purpleidea) [inline resources/aws_ec2.go]: can we flip this and have an early return?
eg req.Method != "POST" { err, return }

File: resources/aws_ec2.go [hunk 4]
```diff
@@ -677,3 +681,40 @@ func (obj *AwsEc2Res) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+	// if addr is a port
+	if _, err := strconv.Atoi(obj.WatchListenAddr); err == nil {
+		addr = fmt.Sprintf(":%s", obj.WatchListenAddr)
+	}
+	srv := &http.Server{Addr: addr}
+	log.Printf("%s: Starting SNS Endpoint", obj)
+	http.HandleFunc("/", obj.postHandler)
+	log.Printf("%s: Listening on %s", obj, addr)
+	obj.wg.Add(1)
+	go func() {
```
Comment (purpleidea) [inline resources/aws_ec2.go]: so this part needs some discussion...
i'm tempted to say that this `run` part of the server should be a different method to keep the awsChan stuff out of this.
eg: this should *only* build the server... meaning we might rename to `snsServer` for example.
perhaps we could merge this without this goroutine until it's clear how best it should get called?
happy to review that as a second patch in this PR if that's what comes next. LMK what you think. (don't just blindly agree with me, disagreeing is fine)

Comment (jonathangold) [reply resources/aws_ec2.go]: I think that's fine. Separation of concerns and all that. I'll remove it from this commit, and add a second one with the listener.

File: resources/aws_ec2.go [hunk 5]
```diff
@@ -677,3 +681,27 @@ func (obj *AwsEc2Res) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+}
+
+// snsPostHandler listens for posts on the SNS Endpoint.
+func (obj *AwsEc2Res) snsPostHandler(w http.ResponseWriter, req *http.Request) {
+	if req.Method != "POST" {
+		http.Error(w, "Invalid request method", http.StatusMethodNotAllowed)
+		return
+	}
+	post, _ := ioutil.ReadAll(req.Body)
+	log.Printf(string(post))
```
Comment (purpleidea) [inline resources/aws_ec2.go]: stale debug message?

File: resources/aws_ec2.go [hunk 6]
```diff
@@ -677,3 +681,27 @@ func (obj *AwsEc2Res) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+
+// snsServer returns an http server used to listen for sns messages.
+func (obj *AwsEc2Res) snsServer() *http.Server {
+	addr := obj.WatchListenAddr
+	// if addr is a port
+	if _, err := strconv.Atoi(obj.WatchListenAddr); err == nil {
+		addr = fmt.Sprintf(":%s", obj.WatchListenAddr)
+	}
+	srv := &http.Server{Addr: addr}
+	log.Printf("%s: Starting SNS Endpoint", obj)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: Are these two log messages still in the same place and/or relevant/useful?

File: resources/aws_ec2.go [hunk 7]
```diff
@@ -677,3 +681,27 @@ func (obj *AwsEc2Res) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
 }
+
+// snsServer returns an http server used to listen for sns messages.
+func (obj *AwsEc2Res) snsServer() *http.Server {
+	addr := obj.WatchListenAddr
+	// if addr is a port
+	if _, err := strconv.Atoi(obj.WatchListenAddr); err == nil {
+		addr = fmt.Sprintf(":%s", obj.WatchListenAddr)
+	}
+	srv := &http.Server{Addr: addr}
```
Comment (purpleidea) [inline resources/aws_ec2.go]: you can return the struct directly, no need for the variable

File: resources/aws_ec2.go [hunk 8]
```diff
@@ -677,3 +681,27 @@ func (obj *AwsEc2Res) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+	log.Printf("%s: Starting SNS Endpoint", obj)
+	http.HandleFunc("/", obj.snsPostHandler)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: weird that this HandleFunc isn't associated with the particular server at all... Is this right? How does this work? What if we had two different http servers running on different ports? isn't that allowed?

Comment (purpleidea) [reply resources/aws_ec2.go]: https://golang.org/src/net/http/server.go?#L2345 is needed AFAICT

File: resources/aws_ec2.go [hunk 9]
```diff
@@ -677,3 +681,30 @@ func (obj *AwsEc2Res) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+	// if addr is a port
+	if _, err := strconv.Atoi(obj.WatchListenAddr); err == nil {
+		addr = fmt.Sprintf(":%s", obj.WatchListenAddr)
+	}
+	handler := http.HandlerFunc(obj.snsPostHandler)
+	return &http.Server{
+		Addr:    addr,
+		Handler: handler,
+	}
+
```
Comment (purpleidea) [inline resources/aws_ec2.go]: nit: kill this extra nl please

File: resources/aws_ec2.go [hunk 10]
```diff
@@ -677,3 +681,30 @@ func (obj *AwsEc2Res) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+// snsPostHandler listens for posts on the SNS Endpoint.
+func (obj *AwsEc2Res) snsPostHandler(w http.ResponseWriter, req *http.Request) {
+	if req.Method != "POST" {
+		http.Error(w, "Invalid request method", http.StatusMethodNotAllowed)
+		return
+	}
+	post, _ := ioutil.ReadAll(req.Body)
+	if obj.debug {
+		log.Printf(string(post))
```
Comment (purpleidea) [inline resources/aws_ec2.go]: do you want to add some sort of prefix to this log message so we know where it came from?

File: resources/aws_ec2.go [hunk 11]
```diff
@@ -71,11 +74,12 @@ var AwsRegions = []string{
...
-	State   string `yaml:"state"`   // state: running, stopped, terminated
-	Region  string `yaml:"region"`  // region must match an element of AwsRegions
-	Type    string `yaml:"type"`    // type of ec2 instance, ie. t2.micro
-	ImageID string `yaml:"imageid"` // imageid must be available on the chosen region
+	BaseRes         `yaml:",inline"`
+	State           string `yaml:"state"`           // state: running, stopped, terminated
+	Region          string `yaml:"region"`          // region must match an element of AwsRegions
+	Type            string `yaml:"type"`            // type of ec2 instance, eg: t2.micro
+	ImageID         string `yaml:"imageid"`         // imageid must be available on the chosen region
+	WatchListenAddr string `yaml:"watchlistenaddr"` // the local address or port that the sns listens on, eg: 10.0.0.0:23456 or 23456
```
Comment (purpleidea) [inline resources/aws_ec2.go]: Let's add one nl before this WatchListenAddr

Might as well keep the basic components in the first chunk. Sorry for the late thought, just trying to think ahead on how to visualize and deal with all the complexity. Thanks!

File: resources/aws_ec2.go [hunk 12]
```diff
@@ -677,3 +682,29 @@ func (obj *AwsEc2Res) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+	}
+}
+
+// snsPostHandler listens for posts on the SNS Endpoint.
+func (obj *AwsEc2Res) snsPostHandler(w http.ResponseWriter, req *http.Request) {
+	if req.Method != "POST" {
+		http.Error(w, "Invalid request method", http.StatusMethodNotAllowed)
+		return
+	}
+	post, _ := ioutil.ReadAll(req.Body)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: why don't we check this error? (sorry)
more importantly, what do we even ever do in this method? it seems basically useless... (or is something coming later?)

Comment (jonathangold) [reply resources/aws_ec2.go]: Yeah. This is just so it has something to do. Once snsSubscribe is done, this will use json.Decoder to read the post, and we can get rid of the ioutil stuff.
